main.go

```diff
@@ -1,732 +1,717 @@
-		for scanner.Scan() {
-			aCachePathItemDef := scanner.Text()
-			aCachePathItemDef = strings.TrimSpace(aCachePathItemDef)
-			if aCachePathItemDef == "" {
-				continue
-			}
-
-			pthItm, err := parseStepParamsPathItemModelFromString(aCachePathItemDef)
-			if err != nil {
-				return StepParamsModel{}, fmt.Errorf("Invalid item: %s", err)
-			}
-
-			stepParams.PathItems = append(stepParams.PathItems, pthItm)
-		}
-
-		fPaths, err := flattenPathItems(stepParams.PathItems)
+	if cacheModel.CompressArchive {
+		gw, err := gzip.NewWriterLevel(tarFile, gzip.BestCompression)
 		if err != nil {
-			return StepParamsModel{}, fmt.Errorf("Failed to filter cache_paths, error: %s", err)
+			return err
 		}
 
-		stepParams.PathItems = fPaths
-
-		if err := scanner.Err(); err != nil {
-			return StepParamsModel{}, fmt.Errorf("Failed to scan the the cache_paths input: %s", err)
-		}
+		cacheModel.GzipWriter = gw
+		cacheModel.TarWriter = tar.NewWriter(gw)
+	} else {
+		cacheModel.TarWriter = tar.NewWriter(tarFile)
 	}
 
-	// Ignore Check on Paths items
-	{
-		scanner := bufio.NewScanner(strings.NewReader(ignoreCheckOnPaths + "\n" + globalIgnoreCheckOnPaths))
-		for scanner.Scan() {
-			aPthItmDef := scanner.Text()
-			aPthItmDef = strings.TrimSpace(aPthItmDef)
-			if aPthItmDef == "" {
-				continue
-			}
-			if strings.HasPrefix(aPthItmDef, "!") {
-				aPthItmDef = strings.TrimPrefix(aPthItmDef, "!")
-				stepParams.IgnorePathsInArchive = append(stepParams.IgnorePathsInArchive, aPthItmDef)
-			}
+	cacheModel.TarFile = tarFile
+	return nil
+}
 
-			stepParams.IgnoreCheckOnPaths = append(stepParams.IgnoreCheckOnPaths, aPthItmDef)
-		}
-		if err := scanner.Err(); err != nil {
-			return StepParamsModel{}, fmt.Errorf("Failed to scan the the cache_paths input: %s", err)
-		}
+// CloseTarArchive ...
+func (cacheModel *CacheModel) CloseTarArchive() error {
+	filePathMapBytes, err := json.Marshal(cacheModel.FilePathMap)
+	if err != nil {
+		return err
 	}
 
-	return stepParams, nil
-}
+	filePathMapSize := int64(len(filePathMapBytes))
 
-func sha1ChecksumOfFile(pth string) ([]byte, error) {
-	fileHasher := sha1.New()
+	closingHeader := &tar.Header{
+		Name:     cacheInfoFilePath,
+		Size:     filePathMapSize,
+		Typeflag: tar.TypeReg,
+		Mode:     0600,
+		ModTime:  time.Now(),
+	}
 
-	f, err := os.Open(pth)
-	if err != nil {
-		return nil, fmt.Errorf("Failed to open file to create checksum, error: %s", err)
+	if err := cacheModel.TarWriter.WriteHeader(closingHeader); err != nil {
+		return err
 	}
-	defer func() {
-		if err := f.Close(); err != nil {
-			log.Printf(" (!) Failed to close file (%s) after creating its checksum, error: %s", pth, err)
-		}
-	}()
 
-	if _, err := io.Copy(fileHasher, f); err != nil {
-		return nil, fmt.Errorf("Failed create hash of file, error: %s", err)
+	if _, err := io.CopyN(cacheModel.TarWriter, bytes.NewReader(filePathMapBytes), filePathMapSize); err != nil && err != io.EOF {
```

Comment:
should not we use: `io.Copy(cacheModel.TarWriter, bytes.NewReader(filePathMapBytes))`, if we copy the whole filePathMapBytes